refactor(op-acceptor): test runner

scharissis opened the pull request:
Refactors the monolithic 1,755-line runner/runner.go file into 5 focused components to improve maintainability and separation of concerns:

  - TestExecutor: Individual test execution and process management
  - OutputParser: Go test JSON output parsing with timeout handling
  - ResultCollector: Hierarchical result aggregation (gates/suites/tests)
  - TestCoordinator: Test workflow orchestration
  - JSONStore: Raw JSON storage and retrieval

scharissis committed:
refactor(op-acceptor): split monolithic runner into focused components

Extract test execution logic into separate components for better maintainability:
- TestExecutor: handles individual test execution
- OutputParser: processes test output parsing
- ResultCollector: aggregates test results
- TestCoordinator: orchestrates test workflows
- JSONStore: manages raw JSON storage

Reduces runner.go complexity while maintaining backward compatibility and test coverage.

serpixel commented on op-acceptor/runner/adapters.go:
```diff
@@ -0,0 +1,106 @@
+	for _, validator := range validators {
+		gateName := validator.Gate
+		if gateName == "" {
+			gateName = "default"
```
Is it possible that we will not have a gate defined? Is it for in-memory tests?

scharissis replied:
Yes.
There's a new "ageless" mode

serpixel commented on op-acceptor/runner/collector.go:
```diff
@@ -0,0 +1,231 @@
+			Description: gateName,
+			Tests:       make(map[string]*types.TestResult),
+			Suites:      make(map[string]*SuiteResult),
+			Status:      types.TestStatusPass,
```
Why are we adding by default the test result as passing?

scharissis replied:
Optimism :)

For simplicity, I didn't want to introduce an 'Unknown' status, which meant this default either had to be Pass,Skip,Fail.
It then gets re-calculated later in `determineSuiteStatus()` and `determineGateStatus()`.

In the interest of time, I'm going to assume that you're concerned that it may result in false positives. I've changed the default to FAIL and added some tests, including to cover the edge case in which we have no tests in a gate.

serpixel commented on op-acceptor/runner/collector.go:
```diff
@@ -0,0 +1,231 @@
+	return c.determineStatusFromFlags(allSkipped, anyFailed)
+}
+
+func (c *resultCollector) determineStatusFromFlags(allSkipped, anyFailed bool) types.TestStatus {
```
I am wondering if we require these methods to be fair, seems like we could simplify this

scharissis replied:
I'm not sure what you mean by these methods being "fair".

I've gone ahead and simplified things by reusing the one package-level implementation.

serpixel commented on op-acceptor/runner/parser.go:
```diff
@@ -0,0 +1,245 @@
+	if hasSkip && result.Status != types.TestStatusFail {
+		result.Status = types.TestStatusSkip
+		if errorMsg.Len() > 0 && result.Error == nil {
+			result.Error = fmt.Errorf("SKIP: %s", strings.TrimSpace(errorMsg.String()))
```
Why are we labelling a skip as an error? Perhaps we just should add it as an info message

scharissis replied:
Good spot, this was a mistake. I've corrected this.

janjakubnanista commented on op-acceptor/runner/adapters.go:
```diff
@@ -0,0 +1,106 @@
+
+// SetConcurrency sets the concurrency level (this is handled at executor creation time)
+func (a *parallelRunnerAdapter) SetConcurrency(concurrency int) {
+	// The existing ParallelExecutor doesn't support runtime concurrency changes
```
Maybe `panic` or at least log here?

scharissis replied:
I've gone ahead and removed this entirely to simplify things.
Cheers.

janjakubnanista commented on op-acceptor/runner/collector.go:
```diff
@@ -0,0 +1,233 @@
+}
+
+// resultCollector implements ResultCollector
+type resultCollector struct{}
```
I'd add a line here to verify that this really implements `ResultsCollector`:

```go
var _ ResultCollector = (*ResultCollector)(nil)
```

janjakubnanista commented on op-acceptor/runner/executor.go:
```diff
@@ -0,0 +1,207 @@
+	ExecutePackage(ctx context.Context, metadata types.ValidatorMetadata) (*types.TestResult, error)
+}
+
+// testExecutor implements TestExecutor
```
Same as above - add that line that confirms it really implements it

janjakubnanista commented on op-acceptor/runner/executor.go:
```diff
@@ -0,0 +1,207 @@
+
+	// Input validation
+	if testDir == "" {
+		panic("testDir cannot be empty")
```
Are we sure we want to `panic` here? Is it too difficult to return and handle errors instead?

scharissis replied:
Agreed. I've refactored this to be error-based.
Cheers

janjakubnanista commented on op-acceptor/runner/jsonstore.go:
```diff
@@ -0,0 +1,64 @@
+)
+
+// jsonStore implements JSONStore interface using FileLogger's RawJSONSink
+type jsonStore struct {
```
Same as above

scharissis replied:
Done

janjakubnanista commented on op-acceptor/runner/parser.go:
```diff
@@ -0,0 +1,244 @@
+)
+
+// outputParser implements OutputParser interface
+type outputParser struct{}
```
Same as above

scharissis replied:
Done

janjakubnanista commented on op-acceptor/runner/parser.go:
```diff
@@ -0,0 +1,244 @@
+
+		if isMainTestEvent(event, metadata.FuncName) {
+			processMainTestEvent(event, result, &testStart, &testEnd, &errorMsg, &hasSkip)
+		} else if event.Test != "" && (strings.Contains(event.Test, "/") || metadata.FuncName == "") {
```
I think it would add tons of readability if this `else` was also pulled out into a well named function, as it is I am struggling to understand what it means (as opposed to `isMainTestEvent` which is nice and understandable)

scharissis replied:
Agreed. Extracted this to a function `isSubTestEvent`.
Cheers

janjakubnanista commented on op-acceptor/runner/runner_test.go:
```diff
@@ -1732,7 +1732,7 @@ gates:
 	skippedTest := result.SubTests["TestSkipped"]
 	require.Equal(t, types.TestStatusSkip, skippedTest.Status, "TestSkipped should have skip status")
-	require.Contains(t, skippedTest.Error.Error(), "This test is intentionally skipped", "Skip message should be preserved")
+	require.Nil(t, skippedTest.Error, "Skipped tests should not have error messages (skip reasons are not errors)")
```
This seems like a big change - going from having error to not having it - are we sure this ain't gonna break some downstreams?

scharissis replied:
It's a change, but not a breaking change, as it's not exposed externally (to consumers).
Good call out, cheers!

janjakubnanista reviewed (changes requested):
Some questions and small stuff, looks good otherwise!

scharissis committed:
chore(op-acceptor): runner; removed redundant function.

Removed 'SetConcurrency()', as it wasn't being used.

scharissis committed:
refactor(op-acceptor): improve error handling in NewTestExecutor

- Replace panics with proper error returns for better error handling
- Add comprehensive unit tests for NewTestExecutor validation
- Remove redundant SetConcurrency method from ParallelRunner interface
- Clean up interface compliance checks in collector

scharissis committed:
refactor(op-acceptor): improve parser conditional readability

- Extract complex subtest logic into isSubTestEvent helper function
- Add comprehensive unit tests for isSubTestEvent with 100% coverage
- Improve code maintainability and documentation
